// @zilliqa-js/zilliqa/dist/index.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@zilliqa-js/core'), require('@zilliqa-js/account'), require('@zilliqa-js/contract'), require('@zilliqa-js/blockchain'), require('@zilliqa-js/subscriptions')) :
  typeof define === 'function' && define.amd ? define(['exports', '@zilliqa-js/core', '@zilliqa-js/account', '@zilliqa-js/contract', '@zilliqa-js/blockchain', '@zilliqa-js/subscriptions'], factory) :
  (factory((global.zilliqa = {}),global['@zilliqa-js/core'],global['@zilliqa-js/account'],global['@zilliqa-js/contract'],global['@zilliqa-js/blockchain'],global['@zilliqa-js/subscriptions']));
}(this, (function (exports,core,account,contract,blockchain,subscriptions) { 'use strict';

  //  This file is part of Zilliqa-Javascript-Library.
  var Zilliqa = /** @class */ (function () {
      function Zilliqa(node, provider) {
          this.provider = provider || new core.HTTPProvider(node);
          this.wallet = new account.Wallet(this.provider);
          this.blockchain = new blockchain.Blockchain(this.provider, this.wallet);
          this.network = new blockchain.Network(this.provider, this.wallet);
          this.contracts = new contract.Contracts(this.provider, this.wallet);
          this.transactions = new account.TransactionFactory(this.provider, this.wallet);
          this.subscriptionBuilder = new subscriptions.SubscriptionBuilder();
      }
      return Zilliqa;
  }());

  exports.Zilliqa = Zilliqa;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
